core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala

```diff
@@ -66,4 +77,68 @@ object DriverWrapper {
         System.exit(-1)
     }
   }
+
+  // R or Python are not supported in cluster mode so download the jars to the driver side
+  private def setupDependencies(loader: MutableURLClassLoader, userJar: String): Unit = {
+    val packagesExclusions = sys.props.get("spark.jars.excludes").orNull
+    val packages = sys.props.get("spark.jars.packages").orNull
+    val repositories = sys.props.get("spark.jars.repositories").orNull
+    val hadoopConf = new HadoopConfiguration()
+    val childClasspath = new ArrayBuffer[String]()
+    val ivyRepoPath = sys.props.get("spark.jars.ivy").orNull
+    var jars = sys.props.get("spark.jars").orNull
+
+    val exclusions: Seq[String] =
+      if (!StringUtils.isBlank(packagesExclusions)) {
+        packagesExclusions.split(",")
+      } else {
+        Nil
+      }
+
+    // Create the IvySettings, either load from file or build defaults
+    val ivySettings = sys.props.get("spark.jars.ivySettings").map { ivySettingsFile =>
+      SparkSubmitUtils.loadIvySettings(ivySettingsFile, Option(repositories),
+        Option(ivyRepoPath))
+    }.getOrElse {
+      SparkSubmitUtils.buildIvySettings(Option(repositories), Option(ivyRepoPath))
+    }
+
+    val resolvedMavenCoordinates = SparkSubmitUtils.resolveMavenCoordinates(packages,
+      ivySettings, exclusions = exclusions)
+
+    if (!StringUtils.isBlank(resolvedMavenCoordinates)) {
+      jars = SparkSubmit.mergeFileLists(jars, resolvedMavenCoordinates)
+    }
+
+    val targetDir = Files.createTempDirectory("tmp").toFile
+    // scalastyle:off runtimeaddshutdownhook
+    Runtime.getRuntime.addShutdownHook(new Thread() {
+      override def run(): Unit = {
+        FileUtils.deleteQuietly(targetDir)
+      }
+    })
+    // scalastyle:on runtimeaddshutdownhook
+
+    val sparkProperties = new mutable.HashMap[String, String]()
+    val securityProperties = List("spark.ssl.fs.trustStore", "spark.ssl.trustStore",
+      "spark.ssl.fs.trustStorePassword", "spark.ssl.trustStorePassword",
+      "spark.ssl.fs.protocol", "spark.ssl.protocol")
+
+    securityProperties
+      .map {pName => sys.props.get(pName)
+        .map{pValue => sparkProperties.put(pName, pValue)}}
+
+    jars = Option(jars).map(SparkSubmit.resolveGlobPaths(_, hadoopConf)).orNull
```

Comment:
These helper methods in `SparkSubmit` should probably be moved to a more properly-named object if they're going to be used in other places. e.g., instead of my previous `IvyUtils` suggestion, you could have a `DependencyUtils` or something like that having all these methods.

Then you could use that method in `DriverRunner` instead of its own `downloadUserJar`.